Add `watchdog_interval` method

Mr0grog opened the pull request:
This adds a new `watchdog_interval` method that returns the current watchdog interval in seconds (or `0` if no interval is set). Fixes #6.

A few notes here:

- I tried to keep this simple by just adding a `watchdog_interval` method and not changing any other code. That said, it might be convenient if `watchdog?` returned `false` if disabled and the interval (instead of `true`) if enabled. That would work fine for most people (since floats are truthy), but might be subtly breaking if someone is literally comparing the return value to `true`.

- Revised the docs for `watchdog?` a bit to hopefully make them clearer for users who aren’t already deeply familiar with `sd_notify`, instead of focusing on under-the-hood bits like environment variables. Happy to change this back if you don’t like that.

- Added tests for `watchdog?` since there weren’t any.

- Added an example of using watchdog to the README. It’s a little complex! Let me know if you have ideas for making it simpler or clearer, or if I should remove it.

- When I ran Rubocop, I got a huge number of errors, mostly from cops that are probably new since this gem last received a major update. I went ahead and turned most of them off in the config, although I did add `required_ruby_version` to the gemspec file since that seemed like a good idea. I set the value (~`>= 2.3.0`~ `>= 2.2.0`) based on the GitHub Actions CI job. Let me know if I should change or remove it.

Mr0grog committed:
Add `watchdog_interval` method and tests

This adds a `watchdog_interval` method as well as tests for it and for `watchdog?` (which previously had no tests). I also took the liberty of revising the docs for `watchdog?` to (I hope) be a little clearer.

Fixes #6.

Mr0grog committed:
Add watchdog example to README

Mr0grog committed:
Update Rubocop config for current Rubocop

Running Rubocop printed out a huge number of new warnings and errors. For the most part, I just silenced them since this gem has been pretty stable and I assume you don't want to modify things just to match newer recommendations. I did add `required_ruby_version` to the gemspec, though, since that just seems like a good idea. It's set to 2.3.0, based on the GitHub Actions CI job.

Mr0grog committed:
Oops, I missed the separate Ruby 2.2 job

Mr0grog committed:
`positive?` does not exist in Ruby 2.2

Mr0grog commented on .rubocop.yml:
```diff
@@ -1,6 +1,9 @@
 require:
   - rubocop-performance
 
+AllCops:
+  TargetRubyVersion: 2.2
+
```
Given how many cops I had to turn *off* in this PR, it might be better to disable them all here and explicitly turn on the ones you want. I’m not sure which you want, though, so I just turned off all the new and/or error-triggering ones.

agis replied:
I guess the issue is that rubocop was never pinned to a specific version. I'd suggest to just disregard all the offenses, given that it's not a requirement in the CI, and just revert this change here, since it's out of context of this PR.

Mr0grog replied:
OK, would it be good to at least remove this Rubocop config file, then? It seems a little confusing to have an unusable `.rubocop.yml` file here.

Mr0grog commented:
Just checking in on this, since it's been sitting for a year. Anything you need from me to land this, @agis?

agis commented on lib/sd_notify.rb:
```diff
@@ -59,33 +59,43 @@ def self.fdstore(unset_env=false)
+    wd_usec = Integer(ENV["WATCHDOG_USEC"])
+    wd_usec > 0 ? wd_usec / 1e6 : 0.0
+  rescue StandardError
+    0.0
```
I think it would be a clearer API if we returned `nil` when the value isn't set. Also I'd prefer to limit this rescue to the place it's intended for, i.e

```ruby
def self.watchdog_interval
  wd_usec = Integer(ENV["WATCHDOG_USEC"]) rescue nil
  wd_usec && wd_usec / 1e6
end
```

WDYT?

Mr0grog replied:
> I think it would be a clearer API if we returned nil when the value isn't set.

Sure. (Just for background, I usually try to avoid returning a different types as a general practice so the JIT can keep methods optimized. This probably isn’t a method that’s going to be on a hot path, though, so that’s not a big deal here.)

agis commented on sd_notify.gemspec:
```diff
@@ -11,6 +11,7 @@ Gem::Specification.new do |s|
   s.files       = ["lib/sd_notify.rb", "LICENSE", "README.md", "CHANGELOG.md"]
   s.homepage    = "https://github.com/agis/ruby-sdnotify"
   s.license     = "MIT"
+  s.required_ruby_version = ">= 2.2.0"
```
What's the rationale behind this requirement?

Mr0grog replied:
See the PR description:

> When I ran Rubocop, I got a huge number of errors, mostly from cops that are probably new since this gem last received a major update. I went ahead and turned most of them off in the config, although I did add `required_ruby_version` to the gemspec file since that seemed like a good idea. I set the value (~`>= 2.3.0`~ `>= 2.2.0`) based on the GitHub Actions CI job. Let me know if I should change or remove it.

That is, Rubocop required that I add `required_ruby_version` or disable that cop. It seemed more reasonable in this case to add it to the gemspec and ask you if you wanted a different value.

From your other comments, it sounds like you'd rather just ignore Rubocop for this repo, though, so I can pull this back out if that’s what you’d like. (Is that what you’d like?)

agis replied:
I see, thanks for elaborating. In that case, yes, let's remove the rubocop config completely.

agis commented on test/sd_notify_test.rb:
```diff
@@ -30,11 +30,50 @@ def test_sd_notify_ready_unset
+    ENV["WATCHDOG_PID"] = $$.to_s
+    setup_socket
+
+    assert_equal(true, SdNotify.watchdog?)
```
This could be simplified to `assert SdNotify.watchdog?

agis reviewed (commented):
Apologies for the radio silence. Left some comments, thanks for this!

agis commented:
Yes, let's remove the rubocop config completely, as you suggested. Thanks

On Sun, Oct 19, 2025, 20:28 Rob Brackett ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sd_notify.gemspec
> <https://github.com/agis/ruby-sdnotify/pull/11#discussion_r2443408982>:
>
> > @@ -11,6 +11,7 @@ Gem::Specification.new do |s|
>    s.files       = ["lib/sd_notify.rb", "LICENSE", "README.md", "CHANGELOG.md"]
>    s.homepage    = "https://github.com/agis/ruby-sdnotify"
>    s.license     = "MIT"
> +  s.required_ruby_version = ">= 2.2.0"
>
> See the PR description:
>
> When I ran Rubocop, I got a huge number of errors, mostly from cops that
> are probably new since this gem last received a major update. I went ahead
> and turned most of them off in the config, although I did add
> required_ruby_version to the gemspec file since that seemed like a good
> idea. I set the value (>= 2.3.0 >= 2.2.0) based on the GitHub Actions CI
> job. Let me know if I should change or remove it.
>
> That is, Rubocop required that I add required_ruby_version or disable
> that cop. It seemed more reasonable in this case to add this to the gemspec
> and ask you if you wanted a different value in this particular case.
>
> From your other comments, it sounds like you'd rather just ignore Rubocop
> for this repo, though, so I can pull this back out if that’s what you’d
> like.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/agis/ruby-sdnotify/pull/11#discussion_r2443408982>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGJ6WCXL2W24DSA4EEJ4HD3YPC5NAVCNFSM6AAAAACGPL5PDKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTGNJUGYYDGNZRHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Mr0grog committed:
Remove Rubocop, per review

The existing Rubocop config is set up for an unknown version of Rubocop from several years ago that was never pinned. The current rules Rubocop suggests are undesired, so this just removes it instead.

Mr0grog committed:
Return `nil` if no interval is set, per review

Mr0grog committed:
Test truthiness, not boolean-ness, per review

Mr0grog committed:
Fix typo in README example

Mr0grog committed:
Use supported runner for old Ruby, test current Ruby

Mr0grog committed:
It turns out we can not longer test on Ruby 2.2

The `ruby/setup-ruby` action does not support Ruby 2.2 on Ubuntu 22+ (see https://github.com/ruby/setup-ruby/issues/496) and GitHub Actions no longer supports older versions of Ubuntu, so we can no longer test Ruby 2.2 (at least not without a lot of extra work!).

Mr0grog committed:
Try an older Minitest version

Mr0grog committed:
Fix test on Ruby 2.1

Mr0grog commented:
@agis OK, I think I’ve made all the changes you asked for and this is ready for re-review.

Since this PR was originally written, [GitHub dropped support for Ubuntu 20.04](https://github.blog/changelog/2025-01-15-github-actions-ubuntu-20-runner-image-brownout-dates-and-other-breaking-changes/), so I had to update the CI config a bit — the “old” tests run in Ubuntu 22.04, and use Ruby 2.1 since [Ruby 2.2 is no longer supported by the `ruby/setup-ruby` action](https://github.com/ruby/setup-ruby/issues/496) on any Ubuntu versions supported on GitHub Actions. I’m assuming based on your previous comments about `required_ruby_version` that you do want to support Ruby 2.1, but I can also just remove the “old” tests entirely if this is not right. (Also: Ruby 2.1 should be installable just fine on Ubuntu 24.04, so we could move all this into the main test job instead having a separate “old” one. The main test job just needs to `bundle exec` so it gets the right minitest version.)

agis commented:
Thanks!
